src/shared/utils/ReactErrorUtils.js

```diff
@@ -12,19 +12,57 @@
 
 'use strict';
 
+var caughtError = null;
+
 var ReactErrorUtils = {
   /**
-   * Creates a guarded version of a function. This is supposed to make debugging
-   * of event handlers easier. To aid debugging with the browser's debugger,
-   * this currently simply returns the original function.
+   * Call a function while guarding against errors that happens within it.
    *
-   * @param {function} func Function to be executed
-   * @param {string} name The name of the guard
-   * @return {function}
+   * @param name (?String) name of the guard to use for logging or debugging
+   * @param func (Function) function to invoke
+   * @param a (*) a First argument
+   * @param b (*) b Second argument
+   */
+  invokeGuardedCallback: function(name, func, a, b) {
+    try {
+      return func(a, b);
+    } catch(x) {
+      if (caughtError === null) {
+        caughtError = x;
+      }
+      return undefined;
+    }
+  },
+
+  /**
+   * During execution of guarded functions we will capture the first error which
+   * we will rethrow to be handled by the top level error handler.
    */
-  guard: function(func, name) {
-    return func;
+  rethrowCaughtError: function() {
+    if (caughtError) {
+      var error = caughtError;
+      caughtError = null;
+      throw error;
+    }
   },
 };
 
+if (__DEV__) {
+  /**
+   * To help development we can get better devtools integration by simulating a
+   * real browser event.
```

Comment:
There is a slight difference in behavior here since this will cause all listeners to be thrown to the top level, instead of just the first one.

That's only really Observable from the debugger though. Potentially you could have an onerror handler that behaves differently but I don't see this as a problem.